File: java/2.x/src/com/relevantcodes/extentreports/ExtentTest.java
/*
* Copyright (c) 2015, Anshoo Arora (Relevant Codes).  All rights reserved.
* 
* Copyrights licensed under the New BSD License.
* 
* See the accompanying LICENSE file for terms.
*/

package com.relevantcodes.extentreports;

import java.io.File;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import com.relevantcodes.extentreports.model.Category;
import com.relevantcodes.extentreports.model.Log;
import com.relevantcodes.extentreports.model.ScreenCapture;
import com.relevantcodes.extentreports.model.Screencast;
import com.relevantcodes.extentreports.model.Test;
import com.relevantcodes.extentreports.model.TestAttribute;
import com.relevantcodes.extentreports.source.ImageHtml;
import com.relevantcodes.extentreports.source.ScreencastHtml;

/** 
 * Defines a node in the report file
 * 
 * @author Anshoo
 *
 */
public class ExtentTest {
    private LogStatus runStatus = LogStatus.UNKNOWN;
    private Test test;
    
    /**
     * Builds a test toggle in the report with the TestName
     * 
     * @param testName Test name
     * @param description A short description of the test
     */
    public ExtentTest(String testName, String description) {
        test = new Test();
        
        test.name = testName == null ? "" : testName.trim(); 
        test.description = description.trim();
        test.startedTime = Calendar.getInstance().getTime();
    }
    
    /**
     * Logs events for the test
     * 
     * @param logStatus Status (see {@link LogStatus})
     * @param stepName Name of the step
     * @param details Details of the step
     */
    public void log(LogStatus logStatus, String stepName, String details) {
        Log evt = new Log();
        
        evt.logStatus = logStatus;
        evt.stepName = stepName == null ? "" : stepName.trim(); 
        evt.details = details == null ? "" : details.trim(); 
        evt.timestamp = Calendar.getInstance().getTime();
                
        test.log.add(evt);
    }
    
    /**
     * Logs events for the test
     * 
     * @param logStatus Status (see {@link LogStatus})
     * @param details Details of the step
     */
    public void log(LogStatus logStatus, String details) {
        log(logStatus, "", details);
    }
    
    /**
     * Allows for adding a snapshot to the log event
     * 
     * @param imgPath Path of the image
     * @return A formed HTML img tag with the supplied path
     */
    public String addScreenCapture(String imgPath) {
        String screenCaptureHtml;
        
        if (isPathRelative(imgPath)) {
            screenCaptureHtml = ImageHtml.getSource(imgPath).replace("file:///", "");
        }
        else {
            screenCaptureHtml = ImageHtml.getSource(imgPath);
        }
        
        ScreenCapture img = new ScreenCapture();
        img.src = screenCaptureHtml;
        img.testName = test.name;
        
        test.screenCapture.add(img);

        return screenCaptureHtml;
    }
    
    /**
     * Allows for adding a screen cast to the log event
     * 
     * @param screencastPath Path of the screencast
     * @return A formed HTML video tag with the supplied path
     */
    public String addScreencast(String screencastPath) {
        String screencastHtml;
        
        if (isPathRelative(screencastPath)) {
            screencastHtml = ScreencastHtml.getSource(screencastPath).replace("file:///", "");
        }
        else {
            screencastHtml = ScreencastHtml.getSource(screencastPath);
        }
        
        Screencast sc = new Screencast();
        sc.src = screencastHtml;
        sc.testName = test.name;
        
        test.screencast.add(sc);
        
        return screencastHtml;
    }
    
    /**
     * Assigns category to test
     * 
     * <p><b>Usage:</b> test.assignCategory("ExtentAPI", "Regression");
     * 
     * @param category Category name
     * @return {@link ExtentTest}
     */
    public ExtentTest assignCategory(String... category) {
        List<String> list = new ArrayList<String>();
        
        for (String c : category) {
            if (!list.contains(c)) {
                test.categoryList.add(new Category(c));
            }
            
            list.add(c);
        }

        return this;
    }
    
    /**
     * Appends a child test to the current test
     * 
     * @param node {@link ExtentTest}
     * @return {@link ExtentTest}
     */
    public ExtentTest appendChild(ExtentTest node) {
        node.getTest().endedTime = Calendar.getInstance().getTime();
        node.getTest().isChildNode = true;
        node.getTest().trackLastRunStatus();
        
        test.hasChildNodes = true;
        
        List<String> list = new ArrayList<String>();
        
        for (TestAttribute attr : this.test.categoryList) {
            if (!list.contains(attr.getName())) {            
                list.add(attr.getName());
            }
        }
        
        for (TestAttribute attr : node.getTest().categoryList) {
            if (!list.contains(attr.getName())) {
                this.test.categoryList.add(attr);
            }
        }
        
        test.nodeList.add(node.getTest());
                
        return this;
    }
    
    /**
     * Provides the current run status of the test
     * 
     * @return {@link LogStatus}
     */
    public LogStatus getRunStatus() {
        return runStatus;
    }
    
    /**
     * Returns the underlying test
     * 
     * @return {@link Test}
     */
    public Test getTest() {        
        return test;
    }
    
    private Boolean isPathRelative(String path) {
        if (path.indexOf("http") == 0 || !new File(path).isAbsolute()) {
            return true;
        }
        
        return false;
    }
}

File: dotnet/ExtentReports/ExtentTest.cs
// ***********************************************************************
// Copyright (c) 2015, Anshoo Arora (Relevant Codes). All rights reserved.
//
// Copyrights licensed under the New BSD License.
//
// See the accompanying LICENSE file for terms.
// ***********************************************************************

namespace RelevantCodes.ExtentReports
{
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Linq;
    using System.Text;

    using Model;
    using Source;

    /// <summary>
    /// Defines a test toggle in the report
    /// </summary>
    public class ExtentTest
    {
        private Test test;

        /// <summary>
        /// Builds a test toggle in the report with the TestName
        /// </summary>
        /// <param name="TestName">Test name</param>
        /// <param name="Description">A short description of the test</param>
        internal ExtentTest(string TestName, string Description)
        {
            test = new Test();

            test.Name = TestName;
            test.Description = Description;
            test.StartedTime = DateTime.Now;
        }

        /// <summary>
        /// Logs events for the test
        /// </summary>
        /// <param name="LogStatus">LogStatus</param>
        /// <param name="StepName">Name of the step</param>
        /// <param name="Details">Step details</param>
        public void Log(LogStatus LogStatus, string StepName, string Details)
        {
            var evt = new Log();

            evt.Timestamp = DateTime.Now;
            evt.LogStatus = LogStatus;
            evt.StepName = StepName;
            evt.Details = Details;

            test.Logs.Add(evt);
        }

        /// <summary>
        /// Logs events for the test
        /// </summary>
        /// <param name="LogStatus">LogStatus</param>
        /// <param name="Details">Step details</param>
        public void Log(LogStatus LogStatus, string Details)
        {
            Log(LogStatus, "", Details);
        }

        /// <summary>
        /// Allows for adding a snapshot to the log event
        /// </summary>
        /// <param name="ImagePath">Path of the image</param>
        /// <returns>A formed HTML img tag with the supplied path</returns>
        public string AddScreenCapture(string ImagePath)
        {
            string screenCaptureHtml;

            if (IsPathRelative(ImagePath))
            {
                screenCaptureHtml = ImageHtml.GetSource(ImagePath).Replace("file:///", "");
            }
            else
            {
                screenCaptureHtml = ImageHtml.GetSource(ImagePath);
            }

            var img = new ScreenCapture();
            img.Source = screenCaptureHtml;
            img.TestName = test.Name;

            test.ScreenCapture.Add(img);

            return screenCaptureHtml;
        }

        /// <summary>
        /// Allows for adding a screen cast to the log event
        /// </summary>
        /// <param name="screencastPath">Path of the screencast</param>
        /// <returns>A formed HTML video tag with the supplied path</returns>
        public string AddScreencast(string screencastPath)
        {
            if (IsPathRelative(screencastPath))
            {
                screencastPath = ScreencastHtml.GetSource(screencastPath).Replace("file:///", "");
            }
            else
            {
                screencastPath = ScreencastHtml.GetSource(screencastPath);
            }

            var sc = new Screencast();
            sc.Source = screencastPath;
            sc.TestName = test.Name;

            test.Screencast.Add(sc);

            return screencastPath;
        }

        /// <summary>
        /// Assigns category to test
        /// </summary>
        /// <param name="CategoryName">Category</param>
        /// <returns>ExtentTest object</returns>
        public ExtentTest AssignCategory(params string[] CategoryName)
        {
            CategoryName
                .ToList()
                .Select(c => c.ToString())
                .Distinct()
                .ToList()
                .ForEach(c => test.CategoryList.Add(new Category(c)));

            return this;
        }

        /// <summary>
        /// Appends a child test to the current test
        /// </summary>
        /// <param name="Node">Child node</param>
        /// <returns>ExtentTest</returns>
        public ExtentTest AppendChild(ExtentTest Node)
        {
            Node.GetTest().EndedTime = DateTime.Now;
            Node.GetTest().IsChildNode = true;
            Node.GetTest().TrackLastRunStatus();
        
            test.HasChildNodes = true;

            IEnumerable<string> testCats = test.CategoryList
                .Select(attr => attr.GetName())
                .Distinct();
            IEnumerable<string> cats = Node.GetTest().CategoryList
                .Select(attr => attr.GetName())
                .Distinct();

            cats.ToList().ForEach(c => { 
                if (!testCats.Contains(c)) 
                    test.CategoryList.Add(new Category(c)); 
            });
        
            test.NodeList.Add(Node.GetTest());

            return this;
        }
        
        internal Test GetTest()
        {
            return test;
        }

        private Boolean IsPathRelative(string FilePath)
        {
            if (FilePath.StartsWith("http") || !Path.IsPathRooted(FilePath))
            {
                return true;
            }

            return false;
        }
    }
}
